cubed/core/array.py

```diff
@@ -186,42 +186,107 @@ def __repr__(self):
+        Parameters
+        ----------
+        work_dir : str or None
+            The directory path (specified as an fsspec URL) used for storing intermediate data.
+        max_mem : int, optional
+            **Deprecated**. The maximum memory available to a worker for data use for the computation, in bytes.
+        allowed_mem : int or str, optional
+            The total memory available to a worker for running a task, in bytes.
+
+            If int it should be >=0. If str it should be of form <value><unit> where unit can be kB, MB, GB, TB etc.
+            This includes any ``reserved_mem`` that has been set.
+        reserved_mem : int or str, optional
+            The memory reserved on a worker for non-data use when running a task, in bytes.
+
+            If int it should be >=0. If str it should be of form <value><unit> where unit can be kB, MB, GB, TB etc.
+        executor : Executor, optional
+            The default executor for running computations.
+        storage_options : dict, optional
+            Storage options to be passed to fsspec.
+        """
 
-    allowed_mem: Optional[int] = None
-    """The total memory available to a worker for running a task, in bytes.
+        if max_mem is not None:
+            warn(
+                "`max_mem` is deprecated, please use `allowed_mem` instead",
+                DeprecationWarning,
+                stacklevel=2,
+            )
 
-    This includes any ``reserved_mem`` that has been set."""
+        self._work_dir = work_dir
 
-    reserved_mem: int = 0
-    """The memory reserved on a worker for non-data use when running a task, in bytes.
+        self._reserved_mem = convert_to_bytes(reserved_mem)
+        if allowed_mem is None:
+            self._allowed_mem = (max_mem or 0) + self.reserved_mem
+        else:
+            self._allowed_mem = convert_to_bytes(allowed_mem)
 
-    See ``cubed.measure_reserved_memory``.
-    """
+        self._executor = executor
+        self._storage_options = storage_options
 
-    executor: Executor = None
-    """The default executor for running computations."""
+    @property
+    def work_dir(self) -> str:
+        """The directory path (specified as an fsspec URL) used for storing intermediate data."""
+        return self._work_dir
 
-    storage_options: dict = None  # type: ignore[assignment]
-    """Storage options to be passed to fsspec"""
+    @property
+    def allowed_mem(self) -> int:
+        """
+        The total memory available to a worker for running a task, in bytes.
 
-    def __post_init__(self):
-        if self.max_mem is not None:
-            warn(
-                "`max_mem` is deprecated, please use `allowed_mem` instead",
-                DeprecationWarning,
-                stacklevel=2,
-            )
-        if self.allowed_mem is None:
-            self.allowed_mem = (self.max_mem or 0) + self.reserved_mem
```

Comment:
Alternatively I could have just replaced this `__post_init__` with just

```python
    def __post_init__(self):
        if self.max_mem is not None:
            warn(
                "`max_mem` is deprecated, please use `allowed_mem` instead",
                DeprecationWarning,
                stacklevel=2,
            )
        object.__setattr__(self, 'reserved_mem', convert_to_bytes(self.reserved_mem))
        if self.allowed_mem is None:
            object.__setattr__(self, 'allowed_mem', (self.max_mem or 0) + self.reserved_mem)
        else:
            object.__setattr__(self, 'allowed_mem', convert_to_bytes(self.allowed_mem))
```

but that's (a) pretty hacky and (b) wouldn't pass a type checker :man_shrugging: